backend/typescript/services/implementations/camperService.ts

```diff
@@ -350,89 +351,120 @@ class CamperService implements ICamperService {
-          existingCampSession,
-          [newWaitlistedCamper],
+      const findCamp = await MgCamp.findById(
+        sessionsToWaitlist[0].camp,
+        {},
+        { session },
+      );
+      if (!findCamp) {
+        throw new Error(
+          "Unable to find the camp associated with the camp sessions passed in",
         );
-      } catch (mongoDbError: unknown) {
-        // rollback user creation
-        try {
-          const deletedWaitlistedCamper: WaitlistedCamper | null = await MgWaitlistedCamper.findByIdAndDelete(
-            newWaitlistedCamper.id,
+      }
+      camp = findCamp;
+
+      // Ensure all camper(s) meet the age requirements of the camp
+      if (
+        !waitlistedCampers.every(
+          (c) => c.age >= camp.ageLower && c.age <= camp.ageUpper,
+        )
+      ) {
+        throw new Error(
+          "Some camper(s) didn't meet the age requirements of the camp",
+        );
+      }
+
+      // Create a waitlist camper entity for each session
+      const campersToWaitlist: Array<Omit<WaitlistedCamperDTO, "id">> = [];
+      campSessions.forEach((cs) => {
+        waitlistedCampers.forEach((waitlistedCamper) => {
+          const camperToWaitlist: Omit<WaitlistedCamperDTO, "id"> = {
+            firstName: waitlistedCamper.firstName,
+            lastName: waitlistedCamper.lastName,
+            age: waitlistedCamper.age,
+            contactName: waitlistedCamper.contactName,
+            contactEmail: waitlistedCamper.contactEmail,
+            contactNumber: waitlistedCamper.contactNumber,
+            campSession: cs,
+            status: "NotRegistered",
+          };
+          campersToWaitlist.push(camperToWaitlist);
+        });
+      });
+      // Insert the waitlisted camper entities
+      newWaitlistedCampers = await MgWaitlistedCamper.insertMany(
+        campersToWaitlist,
+        { session },
+      );
+
+      // Add the campers to the waitlist field of each session
+      await Promise.all(
+        sessionsToWaitlist.map((cs) => {
+          cs.waitlist.push(
+            ...newWaitlistedCampers.filter(
+              (camper) => camper.campSession.toString() === cs.id,
+            ),
           );
+          return cs.save({ session });
+        }),
+      );
 
-          if (!deletedWaitlistedCamper) {
-            throw new Error(
-              `Waitlisted Camper ${newWaitlistedCamper.id} not found.`,
-            );
-          }
-        } catch (rollbackDbError) {
-          const errorMessage = [
-            "Failed to rollback MongoDB waitlisted camper creation after update camp failure. Reason =",
-            getErrorMessage(rollbackDbError),
-            "MongoDB camper id that could not be deleted =",
-            newWaitlistedCamper.id,
-          ];
-          Logger.error(errorMessage.join(" "));
-        }
+      await emailService.sendParentWaitlistConfirmationEmail(
+        camp,
+        sessionsToWaitlist,
+        newWaitlistedCampers,
+      );
```

Comment:
Hmm but what if the confirmation email fails but the transaction is fine? It feels like it'd be better to retry sending the email instead of aborting

Reply:
we could probably discuss sending emails separately? We could send all the emails after the db transactions are completed and then retry sending the email if we fail once. I think that could be it's own ticket though. Could also put that in it's own function and return from the createWaitlistedCamper function once the db transaction is completed

Reply:
Fair enough, do we have a ticket for the fixes backlog?

Reply:
https://www.notion.so/uwblueprintexecs/Task-Board-F22-9e03bf9f7c4343f09a8a5c16dc48a12a?p=bed6cec865e24615b0f6747b6401a2b2&pm=c